NDTensors/src/lib/SymmetrySectors/src/sector_product.jl

```diff
@@ -0,0 +1,241 @@
+# This files defines a structure for Cartesian product of 2 or more fusion sectors
+# e.g. U(1)×U(1), U(1)×SU2(2)×SU(3)
+
+using BlockArrays: blocklengths
+using ..LabelledNumbers: LabelledInteger, label, labelled, unlabel
+using ..GradedAxes: AbstractGradedUnitRange, GradedAxes, dual
+
+# =====================================  Definition  =======================================
+struct SectorProduct{Sectors} <: AbstractSector
+  arguments::Sectors
+  global _SectorProduct(l) = new{typeof(l)}(l)
+end
+
+SectorProduct(c::SectorProduct) = _SectorProduct(arguments(c))
+
+arguments(s::SectorProduct) = s.arguments
+
+# =================================  Sectors interface  ====================================
+function SymmetryStyle(T::Type{<:SectorProduct})
+  return arguments_symmetrystyle(arguments_type(T))
+end
+
+function quantum_dimension(::NotAbelianStyle, s::SectorProduct)
+  return mapreduce(quantum_dimension, *, arguments(s))
+end
+
+# use map instead of broadcast to support both Tuple and NamedTuple
+GradedAxes.dual(s::SectorProduct) = SectorProduct(map(dual, arguments(s)))
+
+function trivial(type::Type{<:SectorProduct})
+  return SectorProduct(arguments_trivial(arguments_type(type)))
+end
+
+# ===================================  Base interface  =====================================
+function Base.:(==)(A::SectorProduct, B::SectorProduct)
+  return arguments_isequal(arguments(A), arguments(B))
+end
+
+function Base.show(io::IO, s::SectorProduct)
+  (length(arguments(s)) < 2) && print(io, "sector")
+  print(io, "(")
+  symbol = ""
+  for p in pairs(arguments(s))
+    print(io, symbol)
+    sector_show(io, p[1], p[2])
+    symbol = " × "
+  end
+  return print(io, ")")
+end
+
+sector_show(io::IO, ::Int, v) = print(io, v)
+sector_show(io::IO, k::Symbol, v) = print(io, "($k=$v,)")
+
+function Base.isless(s1::SectorProduct, s2::SectorProduct)
+  return arguments_isless(arguments(s1), arguments(s2))
+end
+
+# =======================================  shared  =========================================
+# there are 2 implementations for SectorProduct
+# - ordered-like with a Tuple
+# - dictionary-like with a NamedTuple
+
+arguments_type(::Type{<:SectorProduct{T}}) where {T} = T
+
+function sym_arguments_insert_unspecified(s1, s2)
+  return arguments_insert_unspecified(s1, s2), arguments_insert_unspecified(s2, s1)
+end
+
+function arguments_isequal(s1, s2)
+  return ==(sym_arguments_insert_unspecified(s1, s2)...)
+end
+
+# get clean results when mixing implementations
+arguments_isequal(nt::NamedTuple, t::Tuple) = arguments_isequal(t, nt)
+function arguments_isequal(t::Tuple, nt::NamedTuple)
+  isempty(t) && return arguments_isequal((;), nt)
+  isempty(nt) && return arguments_isequal(t, ())
+  return false
+end
```

Comment:
Similar to my suggestion below for `arguments_isless`, we could do something like this:
```julia
function arguments_isequal(a1, a2)
  isempty(a1) && return _arguments_isequal(empty(a2), a2)
  isempty(a2) && return _arguments_isequal(a1, empty(a1))
  return _arguments_isequal(a1, a2)
end

function _arguments_isequal(a1, a2)
  return ==(sym_arguments_insert_unspecified(s1, s2)...)
end
```
`sym_arguments_insert_unspecified/`arguments_insert_unspecified` should catch the case of mixed types and throw an error.

Reply:
`arguments_insert_unspecified` should indeed throw for incompatible types, however `argument_isequal( (U1(1),), (A=SU2(0))` should not throw but return `false`, so I think it should not rely on `arguments_insert_unspecified` for this.

Reply:
Hmm, that's a good point.

Reply:
I guess it could work if we used a function `maybe_sym_arguments_insert_unspecified` that was like `sym_arguments_insert_unspecified` but left the arguments the same as a fallback, i.e.:
```julia
function arguments_isequal(s1, s2)
  isempty(s1) && return _arguments_isequal(empty(s2), s2)
  isempty(s2) && return _arguments_isequal(a1, empty(s1))
  return _arguments_isequal(s1, s2)
end

function _arguments_isequal(s1, s2)
  return ==(maybe_sym_arguments_insert_unspecified(s1, s2)...)
end

function maybe_sym_arguments_insert_unspecified(s1, s2)
  return maybe_arguments_insert_unspecified(s1, s2), maybe_arguments_insert_unspecified(s2, s1)
end

function maybe_arguments_insert_unspecified(t1::Tuple, t2::Tuple)
  n1 = length(t1)
  return (t1..., trivial.(t2[(n1 + 1):end])...)
end

function maybe_arguments_insert_unspecified(nt1::NamedTuple, nt2::NamedTuple)
  diff1 = arguments_trivial(typeof(setdiff_keys(nt2, nt1)))
  return sort_keys(union_keys(nt1, diff1))
end

maybe_arguments_insert_unspecified(s1, s2) = s1
```
I used `maybe_` as a prefix to emphasize the difference from the current code but the name isn't so important.